### PR TITLE
Add Custom Authentication Endpoints.

### PR DESCRIPTION
**Pull request by AbdulrhmnGhanem**
- The endpoint is user/signup, which is equivalent to user/sign_up but returns a JSON response.
- This can't be under the `/api/v1/` route, AFAIK all endpoints under it require being authenticated.
- For the sake of demonstration, not done yet.
- A server configuration is needed.

### CODE REVIEW

**Comment by AbdulrhmnGhanem** [comment]
@kasbah, can you pick the TODOs, please?

**Comment by kasbah** [comment]
What do you mean by "pick the TODOs"?

**Comment by kasbah** [comment]
> The endpoint is user/signup, which is equivalent to user/sign_up

I think that's a confusing route name. Let's call it `user/kitspace_sign_up_api` or something similar.

**Comment by AbdulrhmnGhanem** [comment]
> What do you mean by "pick the TODOs"?

Resolve them.


> > The endpoint is user/signup, which is equivalent to user/sign_up
> 
> I think that's a confusing route name. Let's call it `user/kitspace_sign_up_api` or something similar.

OK.

**Comment by AbdulrhmnGhanem** [comment]
Should I follow this convention `kitsapce_{endpoint}_api` for other endpoints?

**Comment by kasbah** [comment]
I don't have enough context or understanding to work on these TODOs. I think these are firmly within what you are doing. You'll have to explain things better, give me context etc if you want to me to take something on. But it's probably best if the task is something self-contained that I can work on over the course of a few days. 

Yeah keeping the endpoint names consistent makes sense. How many endpoints will you need?

**Comment by AbdulrhmnGhanem** [comment]
> I don't have enough context or understanding to work on these TODOs.

See #5. Otherwise, I do it after finishing the other endpoints. 

> How many endpoints will you need?

* Sign in
* Forgot password
* reset password.

Though I am not sure if anything else will emerge.

File: routers/routes/routes.go [hunk 1]
```diff
@@ -303,6 +303,14 @@ func RegisterRoutes(m *macaron.Macaron) {
...
 	// ***** START: User *****
 	m.Group("/user", func() {
+		m.Group("/kitspace", func() {
+			m.Post("/sign_up", bind(auth.RegisterForm{}), user.KitspaceSignUp)
+			m.Post("/sign_in", bind(auth.SignInForm{}), user.KitspaceSignIn)
+			m.Post("/forgot_password", user.KitspaceForgotPassword)
+			m.Post("/reset_password", user.KitspaceResetPassword)
+			m.Post("/activate", user.KitspaceActivate, reqSignIn)
+		})
+
```
Comment (kasbah) [inline routers/routes/routes.go:313]: :100:

File: routers/user/kitspace_auth.go [hunk 2]
```diff
@@ -50,36 +69,41 @@ func KitspaceSignUp(ctx *context.Context, form auth.RegisterForm) {
...
-	//	response["ActiveCodeLives"] = timeutil.MinutesToFriendly(setting.Service.ActiveCodeLives,
-	//		ctx.Locale.Language())
-	//
-	//	if err := ctx.Cache.Put("MailResendLimit_"+u.LowerName, u.LowerName, 180); err != nil {
-	//		log.Error("Set cache(MailResendLimit) fail: %v", err)
-	//	}
-	//
-	//	ctx.JSON(http.StatusOK, response)
-	//	return
-	//}
```
Comment (kasbah) [inline routers/user/kitspace_auth.go]: Good morning, I think we should still send the confirmation email? I want to test this on staging.kitspace.org later.  I thought you would un-comment this block...

File: routers/user/kitspace_auth.go [hunk 3]
```diff
@@ -70,12 +71,31 @@ func KitspaceSignUp(ctx *context.Context, form auth.RegisterForm) {
...
+			ctx.Locale.Language(),
+		)
+
+		if err := ctx.Cache.Put("MailResendLimit_"+u.LowerName, u.LowerName, 180); err != nil {
+			log.Error("Set cache(MailResendLimit) fail: %v", err)
+		}
+
+		ctx.JSON(http.StatusOK, response)
+	} else {
 		response := make(map[string]bool)
```
Comment (AbdulrhmnGhanem) [inline routers/user/kitspace_auth.go:92]: This else block for local development, I'll remove it once I integrate the frontend forms.

File: routers/user/kitspace_auth.go (hunk 2 above)
Comment (AbdulrhmnGhanem) [reply routers/user/kitspace_auth.go]: Good morning, I re-enabled it!

File: go.mod [hunk 4]
```diff
@@ -23,6 +23,7 @@ require (
 	github.com/alecthomas/chroma v0.8.0
 	github.com/bgentry/speakeasy v0.1.0 // indirect
 	github.com/blevesearch/bleve v1.0.7
+	github.com/cespare/reflex v0.3.0 // indirect
```
Comment (kasbah) [inline go.mod:26]: I think the reflex stuff should be removed?

File: vendor/modules.txt [hunk 5]
```diff
@@ -459,7 +461,7 @@ github.com/jessevdk/go-flags
 ## explicit
 # github.com/json-iterator/go v1.1.10
 github.com/json-iterator/go
-# github.com/kballard/go-shellquote v0.0.0-20170619183022-cd60e84ee657
+# github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
 ## explicit
 github.com/kballard/go-shellquote
 # github.com/kevinburke/ssh_config v0.0.0-20190725054713-01f96b0aa0cd
```
Comment (kasbah) [inline vendor/modules.txt:467]: Should probably not be changed.

File: Dockerfile [hunk 6]
```diff
@@ -19,8 +19,9 @@ COPY . ${GOPATH}/src/code.gitea.io/gitea
 WORKDIR ${GOPATH}/src/code.gitea.io/gitea
 
 #Checkout version if set
-RUN if [ -n "${GITEA_VERSION}" ]; then git checkout "${GITEA_VERSION}"; fi \
- && make clean-all build
+RUN if [ -n "${GITEA_VERSION}" ]; then git checkout "${GITEA_VERSION}"; fi
+RUN make clean-all frontend
+RUN make backend
```
Comment (kasbah) [inline Dockerfile:24]: Should be undone.

**Review by kasbah — CHANGES_REQUESTED** [review]
Can you get this cleaned up for a merge?

File: modules/setting/cors.go [hunk 7]
```diff
@@ -13,13 +13,13 @@ import (
...
-		MaxAge           time.Duration
-		AllowCredentials bool
+		Enabled             bool
+		Scheme              string
+		AllowDomain         []string
+		AllowSubdomain      bool
+		Methods             []string
+		MaxAge              time.Duration
+		AllowCredentials    bool
 	}{
```
Comment (kasbah) [inline modules/setting/cors.go:23]: Formatting changes should be undone.

File: routers/user/kitspace_auth.go (hunk 3 above)
Comment (kasbah) [reply routers/user/kitspace_auth.go]: What's the status of this?

**Comment by kasbah** [comment]
Ready to merge?

**Comment by kasbah** [comment]
Squashed and merged this with https://github.com/kitspace/gitea/commit/3ff28535a127bc9fa9061b2f03a9da7a2d83fb9a

**Pull request closed without merging**